swift/xcode-autobuilder/XcodeProjectParser.cpp

```diff
@@ -213,42 +200,54 @@ static std::unordered_map<std::string, TargetData> mapTargetsToWorkspace(
 static std::vector<fs::path> collectFiles(const std::string& workingDir) {
   fs::path workDir(workingDir);
   std::vector<fs::path> files;
-  auto iterator = fs::recursive_directory_iterator(workDir);
   auto end = fs::recursive_directory_iterator();
-  for (; iterator != end; iterator++) {
-    auto filename = iterator->path().filename();
-    if (filename == "DerivedData" || filename == ".git" || filename == "build") {
-      // Skip these folders
-      iterator.disable_recursion_pending();
+  for (auto it = fs::recursive_directory_iterator(workDir); it != end; ++it) {
+    const auto& p = it->path();
+    if (p.filename() == "Package.swift") {
+      files.push_back(p);
       continue;
     }
-    auto dirEntry = *iterator;
-    if (!dirEntry.is_directory()) {
+    if (!it->is_directory()) {
       continue;
     }
-    if (dirEntry.path().extension() != fs::path(".xcodeproj") &&
-        dirEntry.path().extension() != fs::path(".xcworkspace")) {
+    if (p.filename() == "DerivedData" || p.filename() == ".git" || p.filename() == "build") {
+      // Skip these folders
+      it.disable_recursion_pending();
       continue;
     }
-    files.push_back(dirEntry.path());
+    if (p.extension() == ".xcodeproj" || p.extension() == ".xcworkspace") {
+      files.push_back(p);
+    }
   }
   return files;
 }
 
 static std::unordered_map<std::string, std::vector<std::string>> collectWorkspaces(
-    const std::string& workingDir) {
+    const std::string& workingDir,
+    bool& swiftPackageEncountered) {
   // Here we are collecting list of all workspaces and Xcode projects corresponding to them
   // Projects without workspaces go into the same "empty-workspace" bucket
+  swiftPackageEncountered = false;
   std::unordered_map<std::string, std::vector<std::string>> workspaces;
   std::unordered_set<std::string> projectsBelongingToWorkspace;
   std::vector<fs::path> files = collectFiles(workingDir);
   for (auto& path : files) {
+    std::cerr << path.c_str() << '\n';
     if (path.extension() == ".xcworkspace") {
       auto projects = readProjectsFromWorkspace(path.string());
       for (auto& project : projects) {
         projectsBelongingToWorkspace.insert(project.string());
         workspaces[path.string()].push_back(project.string());
       }
+    } else if (!swiftPackageEncountered && path.filename() == "Package.swift") {
+      // a package manifest must begin with a specific header comment
+      // see https://docs.swift.org/package-manager/PackageDescription/PackageDescription.html
+      static constexpr std::string_view packageHeader = "// swift-tools-version:";
+      char buffer[packageHeader.size()];
+      if (std::ifstream{path}.read(buffer, packageHeader.size()) && buffer == packageHeader) {
+        swiftPackageEncountered = true;
+      }
+      std::cerr << "  " << std::string_view{buffer} << '\n';
```

Comment:
Also left over from printf-debugging?

(FYI: `std::string_view` doesn't have a `char[N]` overload, so it just treats buffer as a null-terminated `char*`, same as streaming `buffer` directly. This will probably change in C++23 with `string_view`'s range constructor.)

Reply:
ah, I expected it to have one, thanks for the info. Anyway these prints will go. They were there while (for some reason I don't understand myself, maybe an autocomplete thing) I put `readsome` instead of `read`, and `readsome` is a strange beast on `ifstream` :sweat_smile: